## SPAM SMS ANALYSIS

#### IMPORT NECESSAREY LIBRARIES

In [38]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import itertools
from collections import Counter
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.preprocessing import StandardScaler
import pingouin as pg
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

In [39]:
df = pd.read_clipboard()
df.head()

v1                                                 v2 Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

### GETTING INFO ABOUT LOADED DATASET

In [40]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   v1          5572 non-null   object
 1   v2          5572 non-null   object
 2   Unnamed: 2  50 non-null     object
 3   Unnamed: 3  12 non-null     object
 4   Unnamed: 4  6 non-null      object
dtypes: object(5)
memory usage: 217.8+ KB


In [41]:
# DESCRIBE

df.describe()

v1                      v2  \
count   5572                    5572   
unique     2                    5169   
top      ham  Sorry, I'll call later   
freq    4825                      30   

                                               Unnamed: 2  \
count                                                  50   
unique                                                 43   
top      bt not his girlfrnd... G o o d n i g h t . . .@"   
freq                                                    3   

                   Unnamed: 3 Unnamed: 4  
count                      12          6  
unique                     10          5  
top      MK17 92H. 450Ppw 16"    GNT:-)"  
freq                        2          2

In [42]:
# SHAPE

df.shape

(5572, 5)

In [43]:
# SIZE
df.size

27860

In [44]:
# Data preprocessing
df['v1'] = df['v1'].map({'ham': 0, 'spam': 1})

In [45]:
# Text preprocessing
tfidf_vectorizer = TfidfVectorizer()
#The TfidfVectorizer in popular machine learning libraries like scikit-learn in Python converts a collection of raw documents 
#into a matrix of TF-IDF features. 
#TF=text frequency
#measure of importance of the term

X = tfidf_vectorizer.fit_transform(df['v2'])
y = df['v1']


In [46]:
# Split the dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [47]:
# Model building
model = LogisticRegression()
model.fit(X_train, y_train)

LogisticRegression()

In [48]:
# Model evaluation
y_pred = model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)

print("Accuracy:", accuracy)
print("Classification Report:\n", report)

Accuracy: 0.9623318385650225
Classification Report:
               precision    recall  f1-score   support

           0       0.96      1.00      0.98       965
           1       1.00      0.72      0.84       150

    accuracy                           0.96      1115
   macro avg       0.98      0.86      0.91      1115
weighted avg       0.96      0.96      0.96      1115



##### This model correctly predicts around 96.23% of the labels in the test set.
##### But this accuracy alone might not be sufficient when dealing with imbalanced datasets.

##### Recall (Sensitivity): Indicates the proportion of actual positives that were correctly identified. 
##### For class 0, it correctly identifies all instances. However, for class 1, it identifies only 72% of the instances.

##### F1-Score: The harmonic mean of precision and recall. It's a balance between precision and recall.
##### For class 0, the F1-Score is high (98%), while for class 1, it's relatively lower (84%).

##### Overall, the model seems to perform well in predicting class 0 (non-spam) but is less effective in identifying class 1 (spam).
##### Improving the model's ability to detect spam (class 1) without compromising the accuracy of non-spam predictions.
##### Might be a focus for further enhancements